Added docs on managing commands through proxies re BDR-4494 (#5172)
EnterpriseDB/docs: product_docs/docs/pgd/5/node_management/maintainance_with_proxies.mdx

## Maintenance commands

When we refer to maintenance commands, we are referring to:

* `VACUUM` 
* Non-replicated DDL commands (which you may want to manually replicate)

## A note on node names

We will be addressing the servers in the cluster by their PGD cluster node names. To get a list of node names in your cluster, use:

```SQL
select node_name from bdr.node;
```

!!! Tip
More details about [`bdr.node`](/pgd/latest/reference/catalogs-visible#bdrnode) table are available in the reference section.
!!!

This will list just the node names. If you need to know which group they are a member of, use:

```
select node_name, node_group_name from bdr.node_summary;
```

!!! Tip
More details about [`bdr.node_summary`](/pgd/latest/reference/catalogs-visible#bdrnode_summary) table are available in the reference section.
!!!

## Using `bdr.run_on_nodes()`
PGD does have the ability to run specific commands on specific nodes using the `bdr.run_on_nodes()` function. This takes two parameters, an array of node names and the command you would like to run on those nodes. For example:

```SQL

> SELECT bdr.run_on_nodes(ARRAY['p-wqnqz5x7ta-a-1','p-wqnqz5x7ta-a-3'],'vacuum full foo');

Author: Will have to go back round and redo all the examples.

__OUTPUT__

                                                            	run_on_nodes
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [{"dsn": "host=p-wqnqz5x7ta-a-1-node.pg.biganimal.io sslmode=verify-full port=5432 dbname=bdrdb", "node_id": "807899305", "response": {"command_status": "VACUUM"}, "node_name": "p-wqnqz5x7ta-a-1", "query_send_time": "2024-01-16 16:24:35.418323+00"}, {"dsn": "host=p-wqnqz5x7ta-a-3-node.pg.biganimal.io sslmode=verify-full port=5432 dbname=bdrdb", "node_id": "199017004", "response": {"command_status": "VACUUM"}, "node_name": "p-wqnqz5x7ta-a-3", "query_send_time": "2024-01-16 16:24:35.4542+00"}]
```

> Would run the vacuum full foo command on the *-1 and *-3 nodes and report the results as JSONB, Note that the node names are passed in an array. The results include the name of the node and the response ( or error message)  resulting from running the command, along with various other fields which may or may not be relevant. It’s also formatted as a single row. To make it more readable, we can apply some formatting.

Reviewer: This paragraph is difficult to digest. 
```suggestion
Would run the vacuum full foo command on the *-1 and *-3 nodes and report the results as a JSON structure. Note that the node names are passed in an array.

The results include the name of the node and the response (or error message) resulting from running the command, along with various other fields. It’s also formatted as a single row.

To make the output more readable, we can apply some formatting.
```

Author: Think its fixed now.

---
title: Maintenance commands through proxies
---

## Maintenance and performance

As a general rule, you should never perform maintenance operations on a cluster's write leader. 
Maintenance operations such as `VACUUM` can be quite disruptive to the smooth running of a busy server and often detrimental to workload performance.
Therefore it is best to run maintenance commands on any node in a group that isn't the write leader.
Generally, this requires you connect directly and issue the maintenance commands on the non-write leader nodes.
But there are situations where this is not possible.

## Maintenance and proxies

Proxies, by design, always connect to and send commands to the current write leader.

> This would usually mean that you should not use a proxy for maintenance.

Reviewer 2: "...should not connect via a proxy to perform maintenance"

Author: Fixed

PGD clusters nodes are able to present a direct connection for psql and pgd cli clients which can be used for issuing maintenance commands to the server on those nodes. 

> But there are environment in which the PGD cluster is deployed wheer a proxy is the only way to access the cluster.

Reviewer 2: typo "wheer"

!!!

This will list just the node names. If you need to know which group they are a member of, use:

```
select node_name, node_group_name from bdr.node_summary;
```

!!! Tip
More details about [`bdr.node_summary`](/pgd/latest/reference/catalogs-visible#bdrnode_summary) table are available in the reference section.
!!!

## Finding the write leader

If we assume that you are connected through the proxy, then you will be connected to the write leader. 
Run `select node_name from bdr.local_node_summary` to see the name of the node:

```
select node_name from bdr.local_node_summary;
__OUTPUT__
    node_name
------------------
node-two
(1 row)
```

This is the node you do **not** want to run your maintenance tasks on. 

```
select * from bdr.node_group_routing_summary;
__OUTPUT__
 node_group_name |    write_lead    | previous_write_lead |       read_nodes
-----------------+------------------+---------------------+-------------------------
 dc1             | node-two         | node-one            | {node-one,node-three}
```

Where the write_lead is the node we determined earlier (node-two), we can also see the two read_nodes (node-one and node-three). 

> It's these nodes that we can safely perform mainten

Reviewer 2: typo: "mainten"
